gemini/sample-apps/llamaindex-rag/src/app/main.py

```diff
@@ -0,0 +1,484 @@
+    answer_similarity,
+    context_precision,
+    context_recall,
+    context_relevancy,
+    faithfulness,
+)
+from src.rag.evaluate import LLMEvaluator, write_results_to_bq
+import uvicorn
+import yaml
+
+logging.basicConfig(
+    filename="eval.log", encoding="utf-8", level=logging.INFO
+)  # Set the desired logging level
+logger = logging.getLogger(__name__)
+
+config_path = os.path.join(
+    os.path.dirname(__file__), "..", "..", "common", "config.yaml"
+)
+with open(config_path, "r") as config_file:
+    config = yaml.safe_load(config_file)
+
+# Now you can use the processed paths
+PROJECT_ID = config["project_id"]
+SERVICE_ACCOUNT_KEY = config["service_account_key"]
+FIRESTORE_DB_NAME = config.get("firestore_db_name")
+
+
+ragas_metrics_dict = {
+    "context_precision": context_precision,
+    "answer_relevancy": answer_relevancy,
+    "faithfulness": faithfulness,
+    "context_relevancy": context_relevancy,
+    "context_recall": context_recall,
+    "answer_similarity": answer_similarity,
+    "answer_correctness": answer_correctness,
+}
+
+
+def get_secret(project_id, secret_id, version_id="latest"):
+    client = secretmanager.SecretManagerServiceClient()
+    name = f"projects/{project_id}/secrets/{secret_id}/versions/{version_id}"
+    response = client.access_secret_version(name=name)
+    return response.payload.data.decode("UTF-8")
+
+
+class IndexUpdate(BaseModel):
+    base_index_name: str
+    base_endpoint_name: str
+    qa_index_name: Optional[str]
+    qa_endpoint_name: Optional[str]
+    firestore_db_name: Optional[str]
+    firestore_namespace: Optional[str]
+
+
+class PromptUpdate(BaseModel):
+    prompt_name: str
+    new_content: str
+
+
+class RAGConfig(BaseModel):
+    llm_name: str = "gemini-1.5-flash"
+    temperature: float = 0.2
+    similarity_top_k: int = 5
+    retrieval_strategy: str = "auto_merging"
+    use_hyde: bool = True
+    use_refine: bool = True
+    use_node_rerank: bool = True
+    use_react: bool = True
+    qa_followup: bool = True
+    hybrid_retrieval: bool = True
+
+
+class RAGRequest(RAGConfig):
+    query: str = "What were Google's Q1 Earnings?"
+    evaluate_response: bool
+    eval_model_name: Optional[str] = "gemini-1.5-flash"
+    embedding_model_name: Optional[str] = "text-embedding-004"
+
+
+class EvalRequest(RAGConfig):
+    eval_model_name: str = "gemini-1.5-flash"
+    embedding_model_name: Optional[str] = "text-embedding-004"
+    input_eval_dataset_bucket_uri: str = "test_rag_questions/test_ground_truth.csv"
+    bq_eval_results_table_id: str = "eval_results.eval_results_table"
+    ragas_metrics: List[str] = ["faithfulness", "answer_relevancy"]
+
+
+# Shared state - to be updated by UI
+from src.app.shared_state import index_manager, prompts
+
+
+def get_index_manager():
+    logger.info(index_manager.base_index.docstore)
+    logger.info(index_manager.firestore_db_name)
+    logger.info(index_manager.firestore_namespace)
+    return index_manager
+
+
+def get_prompts():
```

Comment:
Is there a reason to have this method?

Reply:
Yes, I know it seems odd. The FastAPI backend updates and utilizes state maintained in the `prompts` dataclass. FastAPI implements dependency injection via adding `Depends(<func which returns state>)` so that when a request is made to a particular route, that state is useable within the routes code.